meter.go

```diff
@@ -0,0 +1,269 @@
+are  kbps and pktps for kilobits per second and packets per sec‐
+ond, respectively. The burst option configures the maximum burst
+allowed for the band in kilobits or packets depending on whether
+the unit chosen was kbps or pktps, respectively. If a  burst  is
+not supplied, the switch is free to select some reasonable value
+depending on its configuration.
+
+ovn-nbctl only supports adding a meter with a single  band,  but
+the other commands support meters with multiple bands.
+
+*/
+
+func (odbi *ovndb) rowToMeter(uuid string) *Meter {
+	cacheMeter, ok := odbi.cache[tableMeter][uuid]
+	if !ok {
+		return nil
+	}
+	meter := &Meter{
+		UUID:        uuid,
+		Name:        cacheMeter.Fields["name"].(string),
+		Unit:        cacheMeter.Fields["unit"].(string),
+		Bands:       []string{cacheMeter.Fields["bands"].(libovsdb.UUID).GoUUID},
+		ExternalIds: cacheMeter.Fields["external_ids"].(libovsdb.OvsMap).GoMap,
+	}
+	return meter
+}
+
+func (odbi *ovndb) rowToMeterBand(uuid string) (*MeterBand, error) {
+	cacheMeterBand, ok := odbi.cache[tableMeterBand][uuid]
+	if !ok {
+		return nil, ErrorNotFound
+	}
+	meterBand := &MeterBand{
+		UUID:        uuid,
+		Action:      cacheMeterBand.Fields["action"].(string),
+		Rate:        cacheMeterBand.Fields["rate"].(int),
+		BurstSize:   cacheMeterBand.Fields["burst_size"].(int),
+		ExternalIds: cacheMeterBand.Fields["external_ids"].(libovsdb.OvsMap).GoMap,
+	}
+	return meterBand, nil
+}
+
+/*
+ovn-nbctl only supports adding a meter with a single  band,  but
+the other commands support meters with multiple bands.
+
+meter-add name action rate unit [burst]
+Adds the specified meter. name must be a unique name to identify
+this  meter.  The  action  argument specifies what should happen
+when this meter is exceeded.
+*/
+func (odbi *ovndb) meterAddImp(name, action string, rate int, unit string, external_ids map[string]string, burst ...int) (*OvnCommand, error) {
+
+	//Names  that  start  with "__" (two underscores) are reserved for
+	//internal use by OVN, so ovn-nbctl does not allow adding them.
+	if strings.HasPrefix(name, "__") {
+		return nil, ErrorOption
+	}
+
+	// The only supported action is drop.
+	// If add with wrong option , libovsdb won't accept but doesn't have any feedback. So add some judgment is necessary.
+	if action != "drop" {
+		return nil, ErrorOption
+	}
+
+	MeterUUID, err := newRowUUID()
+	if err != nil {
+		return nil, err
+	}
+	MeterBandUUID, err := newRowUUID()
+	if err != nil {
+		return nil, err
+	}
+	//meter row
+	mRow := make(OVNRow)
+
+	mRow["name"] = name
+	if uuid := odbi.getRowUUID(tableMeter, mRow); len(uuid) > 0 {
+		return nil, ErrorExist
+	}
+
+	mRow["bands"] = libovsdb.UUID{GoUUID: MeterBandUUID}
+
+	switch unit {
+	case "kbps", "pktps":
+		mRow["unit"] = unit
+	default:
+		return nil, ErrorOption
+	}
+
+	//Meter Band row
+	mbRow := make(OVNRow)
+
+	mbRow["action"] = action
+
+	//rate must be in the range 1...4294967295
```

Comment:
but i find it some schema error don't return err after exec , so i think it is a solution

Reply:
Do you have any example? It would be a bug of ovsdb-server if it doesn't enforce the integrity required by schema.

Reply:
like we want to add something like _ at first , thought ovsdb-server make it empty , but it does't return any error,and we get an empty name lr.
{
        "uuid": "02882a77-03b4-4c57-987e-6429458361fe",
        "name": "",
        "ports": [],
        "load_balancer": [],
        "acls": [],
        "qos_rules": [],
        "dns_records": [],
        "other_config": {},
        "external_id": {}
    }

Reply:
What command did you use to add it? Or did you use go-ovn API? I did a quick test with ovn-nbctl ls-add _ls, and there is no problem.

On the other hand, even if it behaves like what you mentioned, it should have nothing to do with schema validation. In ovn-nb schema "name" of logical_switch table is just a "string", without any constraint.

Reply:
yes , i add it use go-ovn api ...and i get an empty name ls

Reply:
i read ovn-nbctl source code and ovn-nbctl just did those validation... like that
https://github.com/ovn-org/ovn/blob/cfb62bb358567405666b6d0c4a6bfa5112f1274c/utilities/ovn-nbctl.c#L2171

Reply:
oh...sorry , the ls-add was my personl problem ..... but validation i think it should be consider.

Reply:
but i successfully add __testMeter1 with go-ovn when I remove the validation
```
__testMeter1: bands:
  drop: 101 kbps, 300 kb burst
testMeter2: bands:
  drop: 101 kbps, 300 kb burst
testMeter3: bands:
  drop: 101 kbps, 300 kb burst
```
which was said it should not add in.
```c
        ctl_error(ctx, "meter names that begin with \"__\" are reserved");
```

Reply:
The schema is strict enough , but ovsdb-server don't return err , when in unit test at go-ovn,we can't get the add-action error until we list them and find nothing.

Reply:
Meter name should not start with "__" is not a constraint defined in schema. In schema file: ovn-nb.ovsschema it is just defined as a string.
        "Meter": {
            "columns": {
                "name": {"type": "string"},
...
That's why it should be checked externally for the __.

For constraints defined in schema, such as range of integer, OVSDB server should detect it and fail the transaction. If the transaction doesn't fail with invalid data, it is a bug either in ovsdb server or in the client (i.e. libovsdb/go-ovn).

It is ok to have validations at the client side even when there is constraint defined in schema, to generate more user-friendly error response, especially for CLIs like ovn-nbctl. But the disadvantage is that the client code is more tightly coupled with the schema definition. It would also be a burden in the client coding if we force the validations for every input at client side. As I said in the beginning, it is not a big deal here.